Implement RRM REST endpoint.

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8791 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -37,7 +71,11 @@ public function set_up() {
 
 		$this->context                = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
-		$this->reader_revenue_manager = new Reader_Revenue_Manager( $this->context );
+		$this->options                = new Options( $this->context );
+		$this->user                   = $this->factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		$this->user_options           = new User_Options( $this->context, $this->user->ID );
```
As these instances are not used outside the constructor, it isn't necessary to define class properties for them. Simple local variables should be sufficient.

```suggestion
		$options                      = new Options( $this->context );
		$user                         = $this->factory()->user->create_and_get( array( 'role' => 'administrator' ) );
		$user_options                 = new User_Options( $this->context, $this->user->ID );
```

nfmohit replied:
As the user information is not used anywhere else, it isn't necessary to define it, `User_Options` will automatically instantiate an instance of it.

```suggestion
		$this->user_options           = new User_Options( $this->context );
```

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -73,4 +111,57 @@ public function test_service_classes_exist() {
 		);
 	}
+
+	public function test_get_data_endpoints() {
```
```suggestion
	public function test_get_datapoints() {
```

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -73,4 +111,57 @@ public function test_service_classes_exist() {
+			function ( Request $request ) {
+				$url = parse_url( $request->getUri() );
+
+				switch ( $url['path'] ) {
+					case '/v1/publications':
+						return new Response( 200 );
+				}
+			}
+		);
+
+		$this->reader_revenue_manager->register();
+
+		$this->authentication->get_oauth_client()->set_granted_scopes(
+			$this->authentication->get_oauth_client()->get_required_scopes()
+		);
+
+		$data = $this->reader_revenue_manager->get_data( 'publications' );
+		$this->assertNotWPError( $data );
+	}
+
+	public function test_service_entity_has_no_access() {
+		FakeHttp::fake_google_http_handler(
+			$this->reader_revenue_manager->get_client(),
+			function ( Request $request ) {
+				$url = parse_url( $request->getUri() );
+
+				switch ( $url['path'] ) {
+					case '/v1/publications':
+						return new Response( 200 );
+				}
+			}
+		);
+
+		$this->reader_revenue_manager->register();
+		$data = $this->reader_revenue_manager->get_data( 'publications' );
+		$this->assertWPError( $data );
+
+		$error_data = $data->get_error_data();
+		$this->assertEquals( 403, $error_data['status'] );
+	}
```
We have built-in test coverage for service entity access, using `Module_With_Service_Entity_ContractTests` trait. Please see its usage in `tests/phpunit/integration/Modules/Analytics_4Test.php` for example.

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
Additional to the test coverage, it would be nice to add test coverage for the `GET:publications` endpoint. This can be done by using a fake HTTP handler and verifying an expected response.

nfmohit replied:
Excellent work here, thanks!

nfmohit commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -51,6 +58,116 @@ public function get_scopes() {
+	/**
+	 * Parses a response for the given datapoint.
+	 *
+	 * @since 1.30.0
```
```suggestion
	 * @since n.e.x.t
```

nfmohit reviewed (changes requested):
Thank you for your excellent work on this, @ankitrox ! Please take at look my comments below:

1. Please address the code review feedback.
2. Please update the QAB.
	1. QA:Eng is not necessary here as the endpoint can simply be tested by using the following script in the browser console in a Site Kit window:
	```js
	googlesitekit.api.get( 'modules', 'reader-revenue-manager', 'publications', undefined, {
	    useCache: false,
	} )
	```
	2. Please add instructions to add publications in the publisher center which can then be seen in the network request response. Multiple publications can be added, some with the URL of the testing site and some without. It should be verified that the response only includes publications with URLs of the testing site.
	
Please let me know if you have any questions, thank you!

ankitrox commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -51,6 +58,116 @@ public function get_scopes() {
+	 * @return boolean|WP_Error
+	 */
+	public function check_service_entity_access() {
+		/**
+		 * Get the subscribe with google service instance.
+		 *
+		 * @var Google_Service_SubscribewithGoogle
+		 */
```
I have added this comment so that editor can direct us to correct method by recognising the object with this comment.

nfmohit replied:
Makes sense 👍 Let's update the comment to enforce naming conventions, e.g. "Google" should be capitalised.

```suggestion
		/**
		 * Get the SubscribewithGoogle service instance.
		 *
		 * @var Google_Service_SubscribewithGoogle
		 */
```

ankitrox commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -51,6 +58,116 @@ public function get_scopes() {
+	protected function create_data_request( Data_Request $data ) {
+		switch ( "{$data->method}:{$data->datapoint}" ) {
+			case 'GET:publications':
+				/**
+				 * Get the subscribe with google service instance.
+				 *
+				 * @var Google_Service_SubscribewithGoogle
+				 */
```
Same as above.

nfmohit replied:
Makes sense 👍 Let's update the comment to enforce naming conventions, e.g. "Google" should be capitalised.

```suggestion
				/**
				 * Get the SubscribewithGoogle service instance.
				 *
				 * @var Google_Service_SubscribewithGoogle
				 */
```

ankitrox commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -67,4 +184,45 @@ protected function setup_info() {
+			);
+		}
+
+		return array(
+			'filter' => $filter,
+		);
```
It can be like this, but we need to pass

```php
[
  'filter' => $filter,
]

```

in `listPublications` method as parameter.

ankitrox commented:
Thank you @nfmohit . 

I've updated the PR as per feedback and left comments on couple of points (regarding var comment).

Assigning over to you for review.

nfmohit commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -55,10 +61,118 @@ public function get_scopes() {
+				)
+			);
+
+			return true;
```
This return statement should be added after the try/catch block, see `Google\Site_Kit\Modules\Analytics_4::check_service_entity_access()` for reference.

nfmohit commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -72,6 +186,47 @@ protected function setup_info() {
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Permutations for site hosts or URL.
```
The function now returns a string.

```suggestion
	 * @return string Permutations for site hosts or URL.
```

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -37,7 +56,13 @@ public function set_up() {
 		parent::set_up();
 
 		$this->context                = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
```
Sorry for not pointing this out earlier, but it looks like a class property isn't required for the context and is not used outside the constructor. A simple local variable should be sufficient.

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -37,7 +56,13 @@ public function set_up() {
+		$this->authentication         = new Authentication( $this->context, $options, $user_options );
+		$this->reader_revenue_manager = new Reader_Revenue_Manager( $this->context, $options, $user_options, $this->authentication );
+
+		$this->enable_feature( 'rrmModule' );
```
The feature flag shouldn't be required for the module internals to work.

```suggestion
```

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -73,4 +98,76 @@ public function test_service_classes_exist() {
+	}
+
+	/**
+	 * @return Reader_Revenue_Manager
```
We usually reference the `Module_With_Service_Entity` class here.

```suggestion
	 * @return Module_With_Service_Entity
```

nfmohit commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -10,12 +10,18 @@
 
-use Google\Site_Kit\Core\Assets\Script;
+use Google\Site_Kit\Core\Authentication\Clients\Google_Site_Kit_Client;
 use Google\Site_Kit\Core\Modules\Module;
 use Google\Site_Kit\Core\Modules\Module_With_Assets;
-use Google\Site_Kit\Core\Modules\Module_With_Assets_Trait;
 use Google\Site_Kit\Core\Modules\Module_With_Scopes;
+use Google\Site_Kit\Core\Modules\Module_With_Assets_Trait;
 use Google\Site_Kit\Core\Modules\Module_With_Scopes_Trait;
+use Google\Site_Kit\Core\Modules\Module_With_Service_Entity;
+use Google\Site_Kit\Core\REST_API\Data_Request;
+use Google\Site_Kit\Core\Util\URL;
+use Google\Site_Kit\Modules\Search_Console\Settings as Search_Console_Settings;
+use Google\Site_Kit_Dependencies\Google\Service\SubscribewithGoogle as Google_Service_SubscribewithGoogle;
```
While we're at it, could we ensure that these are alphabetically ordered? Thanks!

nfmohit commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -11,21 +11,40 @@
 namespace Google\Site_Kit\Tests\Modules;
 
 use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
 use Google\Site_Kit\Modules\Reader_Revenue_Manager;
 use Google\Site_Kit\Tests\TestCase;
+use Google\Site_Kit\Tests\FakeHttp;
+use Google\Site_Kit\Core\Storage\User_Options;
+use Google\Site_Kit\Core\Storage\Options;
+use Google\Site_Kit_Dependencies\GuzzleHttp\Psr7\Request;
+use Google\Site_Kit_Dependencies\GuzzleHttp\Psr7\Response;
+use Google\Site_Kit\Tests\Core\Modules\Module_With_Service_Entity_ContractTests;
+use Google\Site_Kit_Dependencies\Google\Service\SubscribewithGoogle\ListPublicationsResponse;
+use Google\Site_Kit_Dependencies\Google\Service\SubscribewithGoogle\Publication;
```
While we're at it, could we ensure that these are alphabetically ordered? Thanks!

nfmohit reviewed (changes requested):
Thank you for addressing my feedback, @ankitrox ! I've left a few follow-up comments.

Besides, please also take a look at my comment regarding updating the QAB in my last CR feedback.

Thanks!

ankitrox commented:
> Thank you for addressing my feedback, @ankitrox ! I've left a few follow-up comments.
> 
> Besides, please also take a look at my comment regarding updating the QAB in my last CR feedback.
> 
> Thanks!

Thank you @nfmohit for re-review. 

I've updated the PR as per feedback and updated the QAB as well.

nfmohit reviewed (changes requested):
Thank you for all your patience here, @ankitrox ! This looks brilliant now. I've left a couple more very minor comments, so minor that I'll address them myself, I hope that's okay. Thanks!

nfmohit commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -45,7 +52,7 @@ public function register() {
 	 * Gets required Google OAuth scopes for the module.
 	 *
-	 * @since 1.130.0
+	 * @since n.e.x.t
```
This shouldn't have been changed. This methods was already released as part of 1.130.0.

```suggestion
	 * @since 1.130.0
```

nfmohit commented on includes/Modules/Reader_Revenue_Manager.php:
```diff
@@ -55,10 +62,117 @@ public function get_scopes() {
 	 * Sets up information about the module.
 	 *
-	 * @since 1.130.0
+	 * @since n.e.x.t
```
Same here.

```suggestion
	 * @since 1.130.0
```

nfmohit reviewed (changes requested):
@ankitrox , I just found out a few instances of the `@since` tag which were changed to `n.e.x.t` even though their respective methods were already released. I'm updating them as this is a very small change. Let's be mindful of them in the future, thank you!

nfmohit reviewed (approved):
LGTM 👍

aaemnnosttv commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -73,4 +89,76 @@ public function test_service_classes_exist() {
+								'publicationId'         => 'ABCDEFGH',
+								'publicationPredicates' => array(
+									'businessPredicates' => array(
+										'supportsSiteKit' => true,
+										'canSell'         => true,
```
I don't think these exist any more, do they?

aaemnnosttv commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -73,4 +89,76 @@ public function test_service_classes_exist() {
+										'canSell'         => true,
+									),
+								),
+								'verifiedDomains'       => 'example.com',
```
The value here should be an array

aaemnnosttv commented on tests/phpunit/integration/Modules/Reader_Revenue_ManagerTest.php:
```diff
@@ -73,4 +89,76 @@ public function test_service_classes_exist() {
+					case '/v1/publications':
+						$publications = array(
+							array(
+								'publicationId'         => 'ABCDEFGH',
```
I find it's best to create the response using model instances, that way it's less likely to be able to return an invalid shape as we may be here.
